Experimental client: use bundle

jku opened the pull request:
Use the MetadataBundle in the new Updater:
* Remove MetadataWrapper
* Update and expand tests
* Stop using tuf.settings, store the constants in updater_rework.py
* Use MetadataBundle to verify current set of metadata, remove all verification code from Updater
* load local metadata from disk
* persist new verified metadata to disk

Fixes #1304 
Fixes #1320 
Fixes #1319

Commit:
New Updater: use the MetadataBundle

Use the MetadataBundle to verify metadata validity.
* Updater now handles reading metadata files (from filesystem as
  well as network
* Updater feeds bytes to MetadataBundle for verification
* Updater persists data on disk after it had been verified

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
tests: Cleanup in new Updater tests

Add test for a refresh with just a local root.json.

Remove unused code. Add docstrings for raised exceptions, add TODOs for
the missing exception handling.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
The constants could use some thought: We definitely don't want to keep using tuf.settings, but I guess the question is, should these be instance variables in Updater instead of module level ones:

```
MAX_ROOT_ROTATIONS = 32
MAX_DELEGATIONS = 32
DEFAULT_ROOT_MAX_LENGTH = 512000  # bytes
DEFAULT_TIMESTAMP_MAX_LENGTH = 16384  # bytes
DEFAULT_SNAPSHOT_MAX_LENGTH = 2000000  # bytes
DEFAULT_TARGETS_MAX_LENGTH = 5000000  # bytes
```
Maybe it does not hurt if they were instance variables instead. They don't even need to be optional constructor arguments (as initialization does not use any of them: caller can initialize the Updater and then modify the values).
I guess the fanciest option is that constructor takes an optional argument dataclass UpdaterOptions with those attributes ...

jku commented on tuf/client_rework/updater_rework.py:
```diff
@@ -516,7 +331,7 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
+        number_of_delegations = MAX_DELEGATIONS
 
         # Ensure the client has the most up-to-date version of 'targets.json'.
         # Raise 'exceptions.NoWorkingMirrorError' if the changed metadata
```
This is not actually the case: NoWorkingMirrors don't happen anymore, there's TODOs about download errors

sechkova commented on tests/test_updater_rework.py:
```diff
@@ -119,50 +114,25 @@ def setUp(self):
     # Creating a repository instance.  The test cases will use this client
     # updater to refresh metadata, fetch target files, etc.
-    self.repository_updater = updater.Updater(self.repository_name,
+    self.repository_updater = updater.Updater(self.client_directory,
```
`self.repository_name` is not used anymore.

sechkova commented on tests/test_updater_rework.py:
```diff
@@ -187,60 +157,16 @@ def test_refresh(self):
 
     self.assertListEqual(updated_targets, [])
 
+  def test_refresh_with_only_local_root(self):
```
But of course ... files are already copied in the directory. Can we mention this in `test_refresh`, that it actually first loads all metadata files from the file system.

sechkova commented on tuf/client_rework/updater_rework.py:
```diff
@@ -542,14 +357,13 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
             # The metadata for 'role_name' must be downloaded/updated before
             # its targets, delegations, and child roles can be inspected.
-            # self._metadata['current'][role_name] is currently missing.
             # _refresh_targets_metadata() does not refresh 'targets.json', it
```
The rest of the comment is also out-of-date as it references functions not existing in this client version.

sechkova commented on tuf/client_rework/updater_rework.py:
```diff
@@ -610,10 +424,8 @@ def _preorder_depth_first_walk(self, target_filepath) -> Dict:
+                f"{len(role_names)} roles left to visit, but allowed to ",
+                f"visit at most {MAX_DELEGATIONS} delegations.",
             )
             logger.debug(msg)
```
This trick of mine with logging needs to be reverted now that we agreed on %s-strings in the logging calls. 
But not in this PR. I'll note it for myself.

sechkova replied:
Captured in #1400

sechkova commented on tuf/client_rework/updater_rework.py:
```diff
@@ -195,318 +215,113 @@ def download_target(
-
-        return intermediate_targets
+            data = self._load_local_metadata(role)
+            self._bundle.update_delegated_targets(data, role, parent_role)
```
Using `self._bundle.update_delegated_targets` here leaves `MetadataBundle.update_targets()` unused.
Not a big deal but mentioning it.

MVrachev commented:
After I made `consistent_snapshot` optional for root [here](#1394), I realized that the `MetadataBundle` was created before that.
Can we check that if `consistent_snapshot` is not provided everything will work with the update process?

sechkova reviewed (commented):
> The constants could use some thought: We definitely don't want to keep using tuf.settings, but I guess the question is, should these be instance variables in Updater instead of module level ones:
> 
> ```
> MAX_ROOT_ROTATIONS = 32
> MAX_DELEGATIONS = 32
> DEFAULT_ROOT_MAX_LENGTH = 512000  # bytes
> DEFAULT_TIMESTAMP_MAX_LENGTH = 16384  # bytes
> DEFAULT_SNAPSHOT_MAX_LENGTH = 2000000  # bytes
> DEFAULT_TARGETS_MAX_LENGTH = 5000000  # bytes
> ```
> 
> Maybe it does not hurt if they were instance variables instead. They don't even need to be optional constructor arguments (as initialization does not use any of them: caller can initialize the Updater and then modify the values).
> I guess the fanciest option is that constructor takes an optional argument dataclass UpdaterOptions with those attributes ...

Given the proposed structure in https://github.com/theupdateframework/tuf/issues/1397#issuecomment-843103638  where the idea is to expose only the `Updater` class in the client `__init__` , probably the constants need to be inside the class (or the fancier options).  However this can happen when solving #1397. 

I left some minor comments but I in general I like how Updater looks now.

Commit:
New updater: Update comments

Mostly remove comments that provide little value after all the changes.
Also remove a unused variable from a test.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
Addressed the resolved discussions. Also checked that consistent_snapshots=None seems to work fine (using a falsy test)

joshuagl commented on tuf/client_rework/updater_rework.py:
```diff
@@ -195,318 +196,111 @@ def download_target(
+            # Local load failed: we must update from remote
+            logger.debug("Failed to load local snapshot %s", e)
+
+            metainfo = self._bundle.timestamp.signed.meta["snapshot.json"]
+            length = metainfo.get("length") or DEFAULT_SNAPSHOT_MAX_LENGTH
+            version = None
+            if self._bundle.root.signed.consistent_snapshot:
+                version = metainfo["version"]
+
+            data = self._download_metadata("snapshot", length, version)
+            self._bundle.update_snapshot(data)
+            self._persist_metadata("snapshot", data)
```
I was confused at first by the downloading of new metadata all happening in the `except` block. It took me a second read to realise that the `_bundle.update_snapshot()` call fails if the snapshot version is out of date with a `RepositoryError`, and then the logic beneath the `except` clause is executed.

jku replied:
yeah, basically the only case where we don't want to download snapshot is if we have a local snapshot file AND it is valid -- but with exceptions programming this is what it looks like...

Better comments wouldn't hurt

joshuagl commented on tuf/client_rework/updater_rework.py:
```diff
@@ -195,318 +196,111 @@ def download_target(
+                version = metainfo["version"]
+
+            data = self._download_metadata("snapshot", length, version)
+            self._bundle.update_snapshot(data)
+            self._persist_metadata("snapshot", data)
```
Here, and other `_load_*()` functions, the exceptions raised by `MetadataBundle.update_snapshot` and the `open` and `write` calls in `_persist_metadata` become errors the user must handle. The former seems reasonable enough, I'm less certain about the latter. Especially when  `get_one_valid_targetinfo()` -> `_preorder_depth_first_walk()` could result in a myriad of errors (download, file open/write, ???).

Probably a topic for https://github.com/theupdateframework/tuf/issues/1312 ?

jku replied:
yes I vote for handling in 1312.  I think the three classes of exceptions (repository / download / file IO) are logical and also the granularity that I could see a serious client wanting to handle (or not handle as might be reasonable with e.g. file IO ).

joshuagl reviewed (approved):
This looks very concise, nice. It took me a couple of reads to figure out why the fetching of metadata only happened in the exception handler of a load. That might want some attention in terms of comments and/or function names? 🤔 

I also noticed there are some unhandled exceptions which I am not sure whether it is reasonable to leave those to a caller to handle or not. Would welcome thoughts on that.

Overall, very solid. LGTM.
